Refactor run_as into submethods

ViktorHavrysh opened the pull request:
Fixes #19.

Note that 1) I'm pretty new to rust and 2) I haven't run this yet -- need to set up AWS on my home machine...

ViktorHavrysh committed:
Refactor run_as into submethods

jonhoo commented on src/lib.rs:
```diff
@@ -398,36 +409,94 @@ impl TsunamiBuilder {
+                         setup_fns,
+                         usernames,
+                         spot_req_ids,
+                     }| (id_to_name, setup_fns, usernames, spot_req_ids, pk),
```
I'd probably just pass along the entire `SpotRequestResult` here, and then deconstruct in the next future.

jonhoo commented on src/lib.rs:
```diff
@@ -438,338 +507,521 @@ impl TsunamiBuilder {
                 for (name, setup, user, res) in res_v {
-                    setup_fns.insert(name, setup);
-                    usernames.insert(name, user);
+                    result.setup_fns.insert(name, &setup);
```
I don't think the `&` is necessary here?

jonhoo commented on src/lib.rs:
```diff
@@ -438,338 +507,521 @@ impl TsunamiBuilder {
+        ec2: &'a Box<rusoto_ec2::Ec2>,
+        spot_req_ids: Vec<String>,
+        id_to_name: HashMap<String, &'a str>,
+        max_wait: &'a Option<time::Duration>,
```
I think you should just take an `Option<time::Duration>` here. Copying it seems fine.

jonhoo commented on src/lib.rs:
```diff
@@ -438,338 +507,521 @@ impl TsunamiBuilder {
+        // Helper enum for intermediate state
+        enum WaitForInstancesResult<'a> {
+            /// instances, id_to_name
+            Ok(Vec<String>, HashMap<String, &'a str>),
```
Just so the flow is easier to follow, could you make the doc comments:
```rust
enum RunningInstanceSet<'a> {
    /// All instances started correctly: holds instance ids and id_to_name
    All(Vec<String>, HashMap<String, &'a str>),
    /// Some instances were not started: holds instance ids for those that were
    Only(Vec<String>),
}
```

jonhoo commented on src/lib.rs:
```diff
@@ -438,338 +507,521 @@ impl TsunamiBuilder {
+                                        public_ip_address: Some(public_ip),
+                                        ..
+                                    } => {
+                                        let name = id_to_name[&instance_id].clone();
```
I don't think it should be necessary to `clone` here? Just stick to `&*id_to_name[&instance_id]`.

jonhoo commented on src/lib.rs:
```diff
@@ -438,338 +507,521 @@ impl TsunamiBuilder {
+
+                        futures::future::Loop::Break(machines)
+                    })
+                }).map(|machine| (machine, term_instances))
```
Hmm, I don't think this will work? `WaitForInstancesResult::NotOk` would be turned into an `Err` around line 828, which would mean that any instances contained within it would not be mapped up here, which means that they wouldn't be terminated!

ViktorHavrysh replied:
I'm not sure how to resolve this. I'm actually not even sure how it works in the original codebase. I thought I just inlined a closure, but it seems like that changed the behavior somehow?

jonhoo commented on src/lib.rs:
```diff
@@ -438,338 +507,521 @@ impl TsunamiBuilder {
-                        }.map(|_| ())
-                            .map_err(Error::from)
+            .and_then(move |machines| {
+                let machines: HashMap<&str, _> = machines.into_iter().fold(
```
You likely won't need the explicit type annotation here if you make the change to `name` I suggested above.

jonhoo commented on src/lib.rs:
```diff
@@ -438,338 +507,521 @@ impl TsunamiBuilder {
+                let machines: HashMap<&str, _> = machines.into_iter().fold(
+                    HashMap::new(),
+                    |mut machines, (name, machine)| {
+                        machines.entry(&name).or_insert_with(Vec::new).push(machine);
```
The `&` here would also go away.

jonhoo commented on src/lib.rs:
```diff
@@ -438,338 +507,521 @@ impl TsunamiBuilder {
+            })
+            .or_else(move |e| {
+                // 5. terminate all instances
+                if !&term_instances.is_empty() {
```
You shouldn't need the `&` here; `.` automatically does dereferencing if necessary.

jonhoo commented on src/lib.rs:
```diff
@@ -161,6 +161,17 @@ impl MachineSetup {
 }
 
+#[derive(Default)]
+struct SpotRequestResult<'a> {
```
Can you also derive `Debug` and `Clone` here? It's not currently necessary, but those are nice to have :)

ViktorHavrysh replied:
I can't derive debug because the `Fn` doesn't implement it for some reason. I can implement `Clone`, though

jonhoo replied:
Ohh, right, that's too bad. I suppose we could implement `Debug` manually on it, but that should be dealt with in a separate PR if at all :)

jonhoo commented:
This is great! I left some relatively minor comments inline, but those should be relatively straightforward to address :) There is one slightly bigger issue, and that is (as I comment inline) that if only *some* instances start correctly, whereas others do not, those instances will not be terminated (because the `or_else` with the termination code won't be run). In the PR as it stands, termination *only* happens if all the instances start correctly.

ViktorHavrysh committed:
Eliminate most complaints

ViktorHavrysh commented:
I addressed most of your concerns, but I'm not sure what to do with the `term_instances` one. I'm not sure I understand what the original code is doing, to be honest.

ViktorHavrysh commented:
Actually, doesn't the `at_the_end` closure only run in the final `or_else` in the current version of the code as well?

jonhoo commented:
Let me see if I can explain... First, if the timer expires on the setup requests, we produce an `Ok` response that contains the instances that *were* started [here](https://github.com/jonhoo/tsunami/blob/async/src/lib.rs#L701). That means that we *do* enter the `and_then` [here](https://github.com/jonhoo/tsunami/blob/async/src/lib.rs#L742). `at_the_end` is the constructed [here](https://github.com/jonhoo/tsunami/blob/async/src/lib.rs#L746), which means that it will hold the set of instances regardless of whether we timed out or not. Only *after* that do we make the deadline expiring [into an error](https://github.com/jonhoo/tsunami/blob/async/src/lib.rs#L792). This ensures that when `at_the_end` is called, it will know about any running instances regardless of whether the spot requests timed out or not.

And yes, you seem to be totally right! The current `async` code will actually *not* terminate the instances if no error occurred :smile:, because `at_the_end` is *only* invoked in the `or_else` condition [here](https://github.com/jonhoo/tsunami/blob/async/src/lib.rs#L965). It should instead be invoked under a `then` condition so that it always triggers!

jonhoo commented on src/lib.rs:
```diff
@@ -821,14 +820,13 @@ impl TsunamiBuilder {
-                    "not all instances were launched within the time limit"
-                )))
             }
+            RunningInstanceSet::Only(_) => Either::B(futures::future::err(format_err!(
```
Yeah, notice how the instance list is dropped here. It needs to somehow be communicated to the cleanup code so that any instances there will be terminated!

ViktorHavrysh committed:
Refactor error handling, fix problem with instances

ViktorHavrysh commented:
After a long battle with the compiler, I think I fixed the problem with `term_instances`. But I had to refactor error handling in order to do so, and I'm not sure that it turned out better than the initial monolith :)

ViktorHavrysh commented:
Oh, also, in the process of refactoring I turned this:

```rust
match machine
    .public_ip
    .parse::<IpAddr>()
    .context("machine ip is not an ip address")
```

into this:

```rust
let addr = machine
    .public_ip
    .parse::<IpAddr>()
    .expect("machine ip is not an ip address");
```

I the end, this change wasn't really needed, but do we really want to handle the case when an IP is not an IP?

ViktorHavrysh committed:
Improve error handling in case of error during termination

jonhoo commented:
I think this still looks a lot nicer than the old code -- great work! I have some minor comments I'll make inline, but I think we're almost there :D The extra error type is a little unfortunate, but I see why you did it. It seems like a reasonable approach, and no-one has ever complained about *more* expressive errors :p

As for `.expect` on `public_ip`, I agree that it's probably fine to assume that the value will be valid if it is given. I think my concern is more if the value is *empty*, but maybe we're guaranteed that that won't happen? In either case I think the change is okay.

jonhoo commented on src/lib.rs:
```diff
@@ -164,14 +164,72 @@ impl MachineSetup {
+
+    #[fail(display = "tsunami main routine failed")]
+    MainRoutine {
+        error: Box<Error>,
```
I don't think you should need to box this? `failure::Error` is a concrete type, not a trait, and can be stored directly.

jonhoo commented on src/lib.rs:
```diff
@@ -164,14 +164,72 @@ impl MachineSetup {
+enum TsunamiError {
+    #[fail(display = "failed to create security group for new machines: {:?}", error)]
+    SecurityGroupCreate {
+        error: rusoto_ec2::CreateSecurityGroupError,
```
I would make all the things that just have a single error in them be single-element tuples instead so you don't have to write `error:` everywhere :) You [can use `_0`](https://boats.gitlab.io/failure/derive-fail.html#tuple-structs) to refer to the error in `#[fail(display ... )]`.

jonhoo replied:
Another reason to do this is that you can then just do:
```rust
.map_err(TsunamiError::SecurityGroupCreate)
```
Without the extra closure there.

jonhoo commented on src/lib.rs:
```diff
@@ -669,9 +741,9 @@ impl TsunamiBuilder {
+                            return Ok(future::Loop::Continue(id_to_name));
                         } else {
-                            return Err(e).context(format!("failed to describe spot instances"));
+                            return Err(TsunamiError::Describe);
```
I think you're missing the error in `e` here?

jonhoo commented on src/lib.rs:
```diff
@@ -993,49 +1058,56 @@ impl TsunamiBuilder {
-                    .context("failed to clean up key pair")?;
-                    // TODO: clean up created placement group
-                    */
+                ()
```
You shouldn't need the explicit `()` here. Statements that end with `;` evaluate to `()`, so the `info();` above should be sufficient.

jonhoo commented:
It's showing my two last comments as "outdated", but I don't think that's the case..? :/

ViktorHavrysh committed:
Address pull request comments

ViktorHavrysh commented:
Yeah, that was strange. But they should really be outdated now :)

jonhoo commented:
Fantastic! Thank you so much for sticking it out through all my comments. I think the change is looking really good now, so happy to merge :tada: Hope you feel like the experience was helpful too!

ViktorHavrysh commented:
Yeah, I've learned some things, so it's been great :) Hope you'll be continuing your live-coding series.

jonhoo commented:
Also pointed others at this so they can see how we cleaned up the code: https://twitter.com/Jonhoo/status/983024164817833985

EDIT: And yes, the plan is definitely to continue! I have a big deadline coming up in a few weeks, but once I'm past that, I'll be back on schedule :) If you have ideas for things you'd like to see covered, feel free to ping me here, on [Twitter](https://twitter.com/Jonhoo), or on [Patreon](https://www.patreon.com/jonhoo)!

jonhoo commented:
I can also confirm that this now all works with my AWS setup following some [minor changes](https://github.com/jonhoo/tsunami/compare/4860911%5E...async)!

ViktorHavrysh commented:
Rust is the only language I'm familiar with that I'm not very surprised about that fact after that big of a refactor :D
